docs/madr/decisions/056-reachable-svc-mesh-service.md

```diff
@@ -0,0 +1,239 @@
+        //  +optional
+        map<string, string> labels = 5;
+    }
+
+    // List of reachable backend refs via transparent proxy when running with 
+    // MeshExternalService and MeshService. Setting an explicit list
+    // can dramatically improve the performance of the mesh. If not specified,
+    // all services in the mesh are reachable.
+    repeated ReachableBackendRef reachable_backends_ref = 7;
+    }
+```
+
+Most of the fields are optional and may not be used simultaneously. We need to support the following cases:
+* When `name`, `namespace`, and `port` are defined(combination of them, all are not requried), you cannot use `labels`.
+* When `labels` are defined, you cannot use `name`, `namespace`, and `port`.
+
+Another thing that is required to do is add another fields to the `BackendRef` in Dataplane object: namespace and labels
+```proto
+      message BackendRef {
+        // Kind is a type of the object to target. Allowed: MeshService, MeshExternalService
+        string kind = 1;
+        // Name of the targeted object
+        string name = 2;
+        // Port of the targeted object. Required when kind is MeshService, MeshExternalService
+        uint32 port = 3;
+        // Namespace of the targeted object.
+        //  +optional
+        string namespace = 4;
+        // Labels of the targeted object
+        map<string, string> labels = 5;
+      }
+```
+
+This enhancement should make it easier to match the reachable services provided by the user with the generated or provided outbounds.
+
+### Autoreachable services
+
+From a code perspective, we used to build a map of supported tags: `kuma.io/service`, `k8s.kuma.io/namespace`, `k8s.kuma.io/service-name`, and `k8s.kuma.io/service-port`, and later match `MeshTrafficPermission` to dataplanes with these tags. Since the introduction of `MeshService` and `MeshExternalService`, we might not need to iterate over all dataplanes. Instead, we can match all `MeshService` and `MeshExternalService` entries. Based on this information, we know which dataplanes need to be configured. The only limitation is that `MeshService` doesn't carry all tags that a dataplane might have, which means we can support a limited scope of tags, similar to how we do with `kuma.io/service`.
+
+Targetting:
+
+When a user wants to create a `MeshTrafficPermission` that targets a `MeshService`, they cannot do it directly in the top-level TargetRef. Instead, they can use `MeshSubset` to select a specific `MeshService`
+
+```yaml
+kind: MeshService
+spec:
+  selector:
+    dpTags:
+      app: backend
+      k8s.kuma.io/namespace: xyz
+---
+kind: MeshTrafficPermission
+spec:
+  targetRef:
+    kind: MeshSubset
+    tags:
+      app: backend
+```
+
+In the case of `MeshExternalService`, we can introduce a top-level TargetRef for this type, which selects a specific `MeshExternalService`. The problem is that once the resource is synced between zones and a hash is added, the name needs to be the hashed one.
```

Comment:
I'd try to avoid making any decision when it comes to MeshExternalService + Autoreachable services here. We need to first figure out how does MeshExternalService + MeshTrafficPermission work (probably a separate madr). Then we can build autoreachable services on top of that

Reply:
Changed and added a note that we need to decide how to do this. So far we should be fine with `MeshSubset`